### DISCUSSION TITLE
Stateless and Statefull token verification??

### DISCUSSION THREAD

**Discussion by tamis-laan** [lepture/authlib#669] [Q&A]
I used authlib to create `/login` and `/callback` route together with Keycloak. Now I would like to protect my routes. I'm using FastAPI and there is an example in the docs:

https://github.com/authlib/demo-oauth-client/blob/master/fastapi-google-login/app.py 

However I don't want to use a session. I want to verify the `access_token` and the `id_token` for stateless authentication. 

For some more important routes I would like to use statefull verification with keykloak and use the `refresh_token` to issue a new `id_token` and `access_token`.

But I can't find any documentation how to do this properly with Authlib. Any pointers????

**Comment by winwinashwin** [comment]
I stumbled upon the same problem. After going through the Starlette integration code was able to get a stateless version working. If it helps, you can refer to the [implementation here](https://github.com/winwinashwin/fastapi-stateless-oauth2-demo). 

This one encodes the state as JWTs in the redirect request to OAuth server and later in the authorization request decodes it.
